enhancements/installer/openstack-baremetal-compute.md

```diff
@@ -0,0 +1,156 @@
+* Baremetal machines can be added and removed from a running OpenShift cluster.
+
+### Non-Goals
+
+**Separate subnets.**
+
+An OpenStack setting where baremetal machines can't be attached to a virtual
+subnet is not supported in this enhancement.
+
+**External connectivity via floating IPs.**
+
+The OpenStack cluster might make the BM nodes accessible, but exposing them via
+floating IPs might represent an increase in complexity and is currently not a
+goal.
+
+## Proposal
+
+This enhancement is less about adding a new logical piece, than it is about
+uncovering unknowns in a configuration that might work out-of-the-box in
+selected environments.
+
+Therefore, the implementation consists of:
+* identifying a reference architecture
+* anticipating probable issues
+* testing and fixing errors
+
+### Characteristics of the reference architecture
+
+* The choice between deploying a VM or BM node is done by selecting an appropriate Nova flavour
+  * which can be done both in install config and passed to Cluster API provider OpenStack
+  * for the time being, the interactive prompt does ask the flavor for the Compute Nodes specifically; the choice must be made editing `install-config.yaml`
+* The deployment request makes the same API call to Nova as with a VM install (there are no explicit calls to Ironic or anything in the Installer required)
+  * so creating baremetal nodes should work transparently from the installer as well as Cluster API provider OpenStack
+* both VMs and BMs consume the same image from glance
+
+### Probable issues
+
+* The Ignition boot system consisting on a second payload downloaded at runtime
+  requires the baremetal machine to have access to Swift early in the boot
+  process.
+
+* When attaching and detaching, the expected timings must be adjusted to
+  reflect the different boot and shutdown latencies of physical machines
+
+* When attaching, the Certificate Signing Request timeout must be adjusted to
+  reflect the different boot and shutdown latencies of physical machines
+
+### User Stories
+
+#### Increased performance
+
+User wants to run the Compute Nodes on baremetal machines for increased
+performance. The Control Plane sits on regular Nova instances to leverage the
+flexibility of virtual machines.
+
+### Risks and Mitigations
+
+TBD
+
+## Design Details
+
+### Test Plan
+
+The enhancement will be tested using virtual baremetal machines.
+
+TODO: alternatives
+	- OVB on RDO?
+        - MOC is planning on adding Ironic
+	- Install (and snapshot?) an Ironic-enabled OSP on PSI (see Infrared-upshift)
+
+
+TODO: references
+        - https://gitlab.cee.redhat.com/jstransk/infrared-upshift-guide
+	- https://docs.openstack.org/project-deploy-guide/tripleo-docs/latest/environments/virtualbmc.html
+        - https://openstack-virtual-baremetal.readthedocs.io/en/latest/introduction.html
+
+### Graduation Criteria
+
+TBD
+
+##### Removing a deprecated feature
+
+TBD
+
+### Upgrade / Downgrade Strategy
+
+TBD
```

Comment:
Will there be a way to migrate a VM cluster to BM?

That makes me think... what happens if the OpenStack node flavor is modified after the initial deployment? Will the openstack cloud provider detect the change and trigger a re-provision of the nodes?

Reply:
> Will the openstack cloud provider detect the change and trigger a re-provision of the nodes?

Is this a supported feature in OpenShift? It really sounds like it should.

Reply:
I think "migrate" is a key phrase of the question. The openshift model views nodes in the cluster as replaceable, so a migration could be "delete vm, provision baremetal" or the reverse order.

Migrating a control plane via the cloud provider definitely seems like a distinct user story and should be considered out of scope.